### PR TITLE
[DEVOPS-1129] use iohk-nix / nix-tools

### PR DESCRIPTION
**Pull request by jbgi**

### CODE REVIEW

File: nix/.stack.nix/io-sim-classes.nix [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{ system, compiler, flags, pkgs, hsPkgs, pkgconfPkgs, ... }:
+  {
+    flags = {};
+    package = {
+      specVersion = "1.10";
+      identifier = { name = "io-sim-classes"; version = "0.1.0.0"; };
+      license = "NONE";
+      copyright = "2018 IOHK";
+      maintainer = "";
+      author = "Alexander Vieth, Marcin Szamotulski, Duncan Coutts";
```
Comment (coot) [inline nix/.stack.nix/io-sim-classes.nix]: Please also add Karl Knutsson (@karknu) to the authors.

Comment (coot) [reply nix/.stack.nix/io-sim-classes.nix]: Or use `IOHK Engineering Team` as in  `ouroboros-consensus`.

File: nix/.stack.nix/ouroboros-network.nix [hunk 2]
```diff
@@ -0,0 +1,67 @@
+{ system, compiler, flags, pkgs, hsPkgs, pkgconfPkgs, ... }:
+  {
+    flags = {};
+    package = {
+      specVersion = "1.10";
+      identifier = { name = "ouroboros-network"; version = "0.1.0.0"; };
+      license = "NONE";
+      copyright = "2018 IOHK";
+      maintainer = "";
+      author = "Alexander Vieth, Marcin Szamotulski, Duncan Coutts";
```
Comment (coot) [inline nix/.stack.nix/ouroboros-network.nix]: The same here.

**Comment by coot** [comment]
Is `haddock` documentation build along side, is it build on ci?

**Comment by coot** [comment]
Is `haddock` documentation build along side, is it build on ci?

**Comment by jbgi** [comment]
> Is `haddock` documentation build along side, is it build on ci?

No not yet. But soon™: https://github.com/angerman/haskell.nix/pull/13

File: nix/.stack.nix/io-sim-classes.nix (hunk 1 above)
Comment (jbgi) [reply nix/.stack.nix/io-sim-classes.nix]: Everything in `.stack.nix` (and the `.stack-pkgs.nix` file) is generated from cabal/stack config (via `nix/regenerate.sh`).
Probably best to correct cabal files in another PR.

Comment (coot) [reply nix/.stack.nix/io-sim-classes.nix]: ok, we'll do that in another PR

**Comment by coot** [comment]
Am I right that the project is build with `ghc-8.6.3`? We need `ghc-8.4.4` since there is a part that pulls `cardano-sl` as a dependency (in a branch at the moment).

**Comment by jbgi** [comment]
Cross compilation is disabled for now. The `io-sim-classes` does not compile under windows:
```
Configure flags:
--prefix=/nix/store/gnr0mxv2hb3wymm8065w7sy5j1yn20bg-io-sim-classes-0.1.0.0-lib-io-sim-classes-x86_64-pc-mingw32 lib:io-sim-classes --package-db=clear --package-db=/nix/store/06g6xr27qb61mg8mkhl3gnp3n9852ga1-io-sim-classes-0.1.0.0-lib-io-sim-classes-config-x86_64-pc-mingw32/package.conf.d --with-ghc=x86_64-pc-mingw32-ghc --with-ghc-pkg=x86_64-pc-mingw32-ghc-pkg --with-hsc2hs=x86_64-pc-mingw32-hsc2hs --with-gcc=x86_64-pc-mingw32-cc --with-ld=x86_64-pc-mingw32-ld --with-ar=x86_64-pc-mingw32-ar --with-strip=x86_64-pc-mingw32-strip --enable-executable-stripping --enable-library-stripping --hsc2hs-option=--cross-compile --hsc2hs-option=--via-asm 
Configuring library for io-sim-classes-0.1.0.0..
building
Preprocessing library for io-sim-classes-0.1.0.0..
Building library for io-sim-classes-0.1.0.0..
[1 of 6] Compiling Control.Monad.Class.MonadFork ( src/Control/Monad/Class/MonadFork.hs, dist/build/Control/Monad/Class/MonadFork.o )
[2 of 6] Compiling Control.Monad.Class.MonadST ( src/Control/Monad/Class/MonadST.hs, dist/build/Control/Monad/Class/MonadST.o )
[3 of 6] Compiling Control.Monad.Class.MonadSTM ( src/Control/Monad/Class/MonadSTM.hs, dist/build/Control/Monad/Class/MonadSTM.o )
[4 of 6] Compiling Control.Monad.Class.MonadSay ( src/Control/Monad/Class/MonadSay.hs, dist/build/Control/Monad/Class/MonadSay.o )
[5 of 6] Compiling Control.Monad.Class.MonadTimer ( src/Control/Monad/Class/MonadTimer.hs, dist/build/Control/Monad/Class/MonadTimer.o )

src/Control/Monad/Class/MonadTimer.hs:19:1: error:
    Could not find module ‘GHC.Event’
    Use -v to see a list of the files searched for.
   |
19 | import qualified GHC.Event as GHC (TimeoutKey, getSystemTimerManager,
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
builder for '/nix/store/964iq5xmq8259gcl235x6vxh3hgfdmzx-io-sim-classes-0.1.0.0-lib-io-sim-classes-x86_64-pc-mingw32.drv' failed with exit code 1
```
It appears that the `GHC.Event` is not available for Windows. Workaround will need to be implemented for windows support. /cc @edsko

**Comment by edsko** [comment]
You'll have to talk to @dcoutts about the timer stuff. However, on the consensus side, there is also the `IO` layer that needs to be implemented for windows.

**Comment by disassembler** [comment]
bors try

**Comment by disassembler** [comment]
bors ping

**Comment by jbgi** [comment]
bors retry

**Comment by angerman** [comment]
bors retry

**Comment by jbgi** [comment]
bors try

**Comment by cleverca22** [comment]
bors try

**Comment by disassembler** [comment]
bors try

**Comment by jbgi** [comment]
bors try

**Comment by jbgi** [comment]
bors try

**Comment by jbgi** [comment]
bors try

**Comment by jbgi** [comment]
bors try

**Comment by rvl** [comment]
The FS tests are failing on macOS? `RemoveFile (PExpPath []) => FsResourceInappropriateType`.

**Comment by mrBliss** [comment]
> The FS tests are failing on macOS? `RemoveFile (PExpPath []) => FsResourceInappropriateType`.

Raised in #208 and fixed in #210.

**Comment by coot** [comment]
Indeed, we are targeting 8.6.3 but we also need to build `cardano-sl` alongside. @avieth could you check if this will work for you.

**Comment by coot** [comment]
When I run 
```
nix-build ./default.nix -A nix-tools.libs.typed-protocols  
trace: DEPRECATED: use mkPkgSet instead of mkNewPkgSet
error: attribute 'typed-protocols' in selection path 'nix-tools.libs.typed-protocols' not found
```

What is the way to build `typed-protocols`? This worked for `ouroboros-network` and other cabal packages defined in this repo; maybe it's missing somewhere.

**Comment by jbgi** [comment]
@coot isn't `typed-transitions` instead of `typed-protocols`?

**Comment by coot** [comment]
@jbgi it's both at the moment.

**Comment by jbgi** [comment]
@coot ah, yes, I see #209 which introduce `typed-protocols` but not merged so not available in this PR.
`typed-transitions` is available though:
`nix build -f default.nix nix-tools.libs.typed-transitions` 
or
`nix-build default.nix -A nix-tools.libs.typed-transitions`
should work.

File: default.nix [hunk 3]
```diff
@@ -1,22 +1,7 @@
-# Builds using system <nixpkgs> and ghc863 with the required package curation.
```
[the remark this answers is not in the document]
Comment (jbgi) [reply default.nix]: @coot Added more documentation. Let me know if I missed something or if unclear.

File: release.nix [hunk 4]
```diff
@@ -1,17 +1,50 @@
...
 
-    platforms = {
-      typed-transitions = supportedSystems;
-      ouroboros-network = supportedSystems;
-      ouroboros-consensus = supportedSystemd;
-    };
-in rlib.mapTestOn platforms // ouroboros-network
+  # packages from our stack.yaml or plan file (via nix/pkgs.nix) we
+  # are interested in building on CI via nix-tools.
+  packages = [ "ouroboros-consensus" "ouroboros-network" ];
```
[the remark this answers is not in the document]
Comment (coot) [reply release.nix]: @jbgi I think we should also add the above list of packages

**Comment by coot** [comment]
@jbgi can we merge it, or do you want to add something on top?

**Comment by jbgi** [comment]
bors try

**Comment by coot** [comment]
@jbgi this PR needs a rebase, can we merge it after resolving the conflicts?

**Comment by jbgi** [comment]
bors try

**Pull request merged**